File: Design Patterns/Strategy/strategy.cpp
#include <iostream>
using namespace std;

///////////////////////////////////////////
// Strategy Interface: DriveStrategy
///////////////////////////////////////////

class DriveStrategy
{
public:
  virtual void drive() = 0; // Pure virtual function
};

///////////////////////////////////////////
// Concrete Strategies
///////////////////////////////////////////

class NormalDriveStrategy : public DriveStrategy
{
public:
  void drive() override
  {
    cout << "Normal drive capability" << endl;
  }
};

class SportsDriveStrategy : public DriveStrategy
{
public:
  void drive() override
  {
    cout << "Sports drive capability" << endl;
  }
};

class XYZDriveStrategy : public DriveStrategy
{
public:
  void drive() override
  {
    cout << "XYZ drive capability" << endl;
  }
};

///////////////////////////////////////////
// Vehicle Base Class (Context)
///////////////////////////////////////////

class Vehicle
{
private:
  DriveStrategy *driveObj; // Strategy object

public:
  // Constructor injection
  Vehicle(DriveStrategy *driveObj)
  {
    this->driveObj = driveObj;
  }

  virtual void drive()
  {
    driveObj->drive();
  }

  // Virtual destructor to avoid memory leaks
  virtual ~Vehicle()
  {
    delete driveObj;
  }
};

///////////////////////////////////////////
// Concrete Vehicle Types
///////////////////////////////////////////

class SportyVehicle : public Vehicle
{
public:
  SportyVehicle() : Vehicle(new SportsDriveStrategy()) {}
};

class PassengerVehicle : public Vehicle
{
public:
  PassengerVehicle() : Vehicle(new NormalDriveStrategy()) {}
};

class OffRoadVehicle : public Vehicle
{
public:
  OffRoadVehicle() : Vehicle(new XYZDriveStrategy()) {}
};

///////////////////////////////////////////
// Main Function
///////////////////////////////////////////

int main()
{
  Vehicle *vehicle = new SportyVehicle();
  vehicle->drive();

  delete vehicle;
  return 0;
}

File: Design Patterns/Behavioural/Strategy/strategy.py
from abc import ABC, abstractmethod

#########################################
# Strategy Interface
#########################################

class DriveStrategy(ABC):
    @abstractmethod
    def drive(self):
        pass

#########################################
# Concrete Strategies
#########################################

class NormalDriveStrategy(DriveStrategy):
    def drive(self):
        print("Normal drive capability")

class SportsDriveStrategy(DriveStrategy):
    def drive(self):
        print("Sports drive capability")

class XYZDriveStrategy(DriveStrategy):
    def drive(self):
        print("XYZ drive capability")

#########################################
# Vehicle Base Class (Context)
#########################################

class Vehicle:
    def __init__(self, drive_strategy: DriveStrategy):
        self.drive_strategy = drive_strategy  # Strategy injection

    def drive(self):
        self.drive_strategy.drive()

#########################################
# Concrete Vehicle Types
#########################################

class SportyVehicle(Vehicle):
    def __init__(self):
        super().__init__(SportsDriveStrategy())

class PassengerVehicle(Vehicle):
    def __init__(self):
        super().__init__(NormalDriveStrategy())

class OffRoadVehicle(Vehicle):
    def __init__(self):
        super().__init__(XYZDriveStrategy())

#########################################
# Main Function
#########################################

if __name__ == "__main__":
    vehicle = SportyVehicle()
    vehicle.drive()
